src/Firestore/Query.php

```diff
@@ -0,0 +1,477 @@
+        }
+    }
+
+    /**
+     * Get all documents matching the provided query filters.
+     *
+     * This method will not execute an immediate service call. The Query will be
+     * executed once you begin iterating over the rows (either by iterating
+     * directly on the QuerySnapshot, or by iterating over
+     * {@see Google\Cloud\Firestore\QuerySnapshot::rows()}).
+     *
+     * Example:
+     * ```
+     * $result = $query->snapshot();
+     * ```
+     *
+     * @param array $options Configuration options.
+     * @return QuerySnapshot
+     */
+    public function snapshot(array $options = [])
+    {
+        $options['maxRetries'] = (isset($options['maxRetries']))
+            ? $options['maxReties']
+            : FirestoreClient::MAX_RETRIES;
+
+        $call = function () use ($options) {
+            unset($options['maxRetries']);
+
+            return $this->connection->runQuery([
+                'parent' => $this->parent,
+                'structuredQuery' => $this->query,
+                'transaction' => $this->transaction,
+                'retries' => 0,
+            ] + $options);
+        };
+
+        return new QuerySnapshot($this->connection, $this->valueMapper, $this, $call, $options['maxRetries']);
+    }
+
+    /**
+     * Add a SELECT to the Query.
+     *
+     * Example:
+     * ```
+     * $query = $query->select(['firstName']);
+     * ```
+     *
+     * @param array $fieldPaths The projection to return, in the form of an
+     *        array of field paths. To only return the name of the document, use
+     *        `['__name__']`.
+     * @return Query A new instance of Query with the given changes applied.
+     */
+    public function select(array $fieldPaths)
+    {
+        $fields = [];
+        foreach ($fieldPaths as $field) {
+            $fields[] = [
+                'fieldPath' => $field
+            ];
+        }
+
+        return $this->newQuery([
+            'select' => [
+                'fields' => $fields
+            ]
+        ]);
+    }
+
+    /**
+     * Add a WHERE clause to the Query.
+     *
+     * Example:
+     * ```
+     * $query = $query->where('firstName', '=', 'John');
+     * ```
+     *
+     * @param string $fieldPath The field to filter by.
+     * @param string $operator The operator to filter by.
+     * @param mixed $value The value to compare to.
+     * @return Query A new instance of Query with the given changes applied.
+     */
+    public function where($fieldPath, $operator, $value)
+    {
+        $operator = array_key_exists($operator, $this->shortOperators)
+            ? $this->shortOperators[$operator]
+            : $operator;
+
+        if (!in_array($operator, $this->allowedOperators)) {
+            throw new \BadMethodCallException(sprintf(
+                'Operator %s is not a valid operator',
+                $operator
+            ));
+        }
+
+        $query = [
+            'where' => [
+                'compositeFilter' => [
```

Comment:
Do you support UnaryFilters on NaN and Null?

Reply:
this is added now.